### PR TITLE
Add a new Woo-branded flow for connecting a Jetpack site

### PR DESCRIPTION
**Pull request by DanReyLop**
See paJDYF-Ei-p2 for more context and design of the actual flow.

The intention with this flow is to provide an exclusively Woo-branded flow for connecting a site to Jetpack. It will be used in combination with the `jetpack-connection` package bundled in the WooCommerce Payments, so when the user starts going through the flow he may have never read the word "Jetpack". I've named the flow "Woo DNA", because Woo + Jetpack DNA. I'm not very creative.

I've divided this PR into separate commits as best I could, so that's the most sensible way tor review it.

In https://github.com/Automattic/wp-calypso/commit/4164b47a60d19944f62d3edf5bc208300598de6d, the "flow configuration" is added. Right now, WooCommerce Payments is the only Woo extension that will use it, but more plugins can use the Woo flow if they are added to that list.

### To test:

Checkout this branch of `wp-calypso` and run it locally (`yarn && yarn start`, add `127.0.0.1 calypso.localhost` to your `/etc/hosts`).

Add these constants to your local WordPress site:
```
define( 'WOOCOMMERCE_CALYPSO_ENVIRONMENT', 'development' );
define( 'JETPACK__SANDBOX_DOMAIN', 'yourtestblog.wordpress.com' );
```
`yourtestblog.wordpress.com` needs to be a WordPress.com site that has its DNS pointing to your wp.com sandbox. See here how I did it: pMz3w-bfP-p2

~You'll need to add `yourtestblog.wordpress.com <YOUR_SANDBOX_IP` to `/etc/hosts`. Alternatively, you can use my test blog so you don't need to make a Systems Request. Ping me to give you my blog domain and IP address.~ This is not needed.

Follow the instructions in https://github.com/Automattic/woocommerce-payments/pull/638 to start a Jetpack connection from WooCommerce Payments.

Hopefully, you'll be redirected to `calypso.localhost:3000/...`, and the flow (behaviour, color scheme, text) will match exactly the designs on paJDYF-Ei-p2 . Ping me if you can't get it up and running.

Note that all these PRs need to be merged first, as they have tweaks and refactors needed for this one to work: #41799 #41800 #41801 #41802 #41803 #41804 #41805

### i18n

Here are all the new strings to translate. The `Woocommerce logo` one is the `alt` attribute of the Woo logo that's in the header.

Screenshot 2020-06-13 at 19 00 25
Screenshot 2020-06-13 at 19 01 45
Screenshot 2020-06-13 at 19 02 13

### CODE REVIEW

File: client/jetpack-connect/signup.js
Comment (a8ci18n) [inline client/jetpack-connect/signup.js:285]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Confirm Login to WordPress.com' )` *ES Score: 7*

**Comment by v18** [comment]
@DanReyLop - is the login screen supposed to be Woo-branded already?

**Comment by DanReyLop** [comment]
@v18 Yes, everything should be Woo-branded

**Comment by v18** [comment]
@DanReyLop I'm still seeing a JP-branded login screen. I haven't dug in to find the root cause of this yet (so, it could be my setup potentially), but I wanted to let you know while I review the other parts:

Screen Shot 2020-05-27 at 6 25 12 PM

Steps to see that screen:
1. Make sure you are logged out of wp.com
2. Go to the Payments tab in your wp-admin
3. Click the 'Set up' button to start the connection
4. See that page

File: config/development.json
Comment (v18) [inline config/development.json:85]: Having a hard time finding this in the docs - which enviornment does this turn it on for? Only `calypso.localhost:3000`?

Comment (DanReyLop) [reply config/development.json:85]: Yes, only locally for now.

File: client/layout/logged-out.jsx
Comment (v18) [inline client/layout/logged-out.jsx:157]: I'd expect this variable to be a boolean, not an object or undefined. Should we wrap this in `Boolean()`?

File: config/development.json
Comment (v18) [reply config/development.json:85]: What's the plan for turning the feature on for staging & production?

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:331]: Why do we need two separate if blocks checking for `this.state.showWooDnaLoginForm`?

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:273]: Here we're either showing the log in form or the sign up form. I think this could use a comment at the top of the if block to explain.

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:73]: This param is being used to either show the log in form or the sign up for, but specifically only if it's a "woo dna" flow. Have you considered having it store the type of form we want to display instead of a boolean? For example either "signup" or "login", etc?

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:126]: What does `resetAuthAccountType` do in this case?

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:401]: What's the significance of the auth account type to whether or not the full login form is visible?

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:110]: How did the original flow handle getting an unknown using when logging in?

File: client/jetpack-connect/signup.js
Comment (v18) [inline client/jetpack-connect/signup.js:274]: This is basically the difference between the "enter your email" screen and the "log in" screen right? Maybe we could either add a comment or change the variable name to something like `! isLoggedOutScreen`. Or we could designate three states in the component local state for the type of Woo DNA screen we're on - "log in", "sign up" or "logged out" (or "ask for email"). I commented about having a non-binary state variable this like this in another comment, too.

File: client/jetpack-connect/authorize.js
Comment (v18) [inline client/jetpack-connect/authorize.js:761]: Nitpick: could this be `wooDna={ wooDna }`?

File: client/jetpack-connect/woo-dna-config.js
Comment (v18) [inline client/jetpack-connect/woo-dna-config.js:1]: I think we could add some comments to this file explaining how to extend the usage for new woo flows for later. Here are two comments I'd like to see:

1. One thing we could explain is that the `from` parameter is checked against the params of the object - like "woocommerce-payments". And that we may want to consider changing the structure to include a parameter to check for instead of checking the name in the future, if more complex checks need to be done.

2. Another thing I think should be explained is the difference between a "woo Dna" flow and the other woo flows that exist.

File: client/blocks/login/login-form.jsx
Comment (v18) [inline client/blocks/login/login-form.jsx:486]: The variables `isJetpackWooCommerceFlow` and `isJetpackWooDnaFlow` are really similar and it's hard to understand how they're different to someone new to this code. I'm trying to think how to make it clearer what's going on.

The biggest difference is that one flow involves a "Jetpack" branded flow that is different from the default Jetpack flow.

The "DNA" flow is using the Jetpack Connection package. This flow is essentially a "WP.com" connection flow.

Is DNA an official name that's being used? I don't see it in the JP connection package docs:

https://github.com/Automattic/jetpack/tree/master/packages/connection

What do you think? I couldn't really come up with anything better. I had some ideas like `isWooWPcomConnectionFlow` but that's not really clear either. Happy to leave it as is, especially if we add some comments in the config file, but wanted to mention in case you have a new idea as you're reading this.

**Comment by v18** [comment]
Great job @DanReyLop . This PR works really nicely. Left some comments on the code, and I have a few questions about the overall PR:
- are we recording tracks events separately for this flow? any events that are tracked as part of this flow that were already being tracked before these PRs?
- should we start a new issue for the Jetpack-branded screen showing up, or is that going to be solved as part of this PR?
- for the "to do" items mentioned in the PR description (passwordless flow, etc) - should we record those tasks in a separate issue to make sure we can keep track?

File: config/development.json
Comment (DanReyLop) [reply config/development.json:85]: We'll first merge this into `master` as-is, and wait a bit in case there are regressions. Before the "call for testing", we'll turn the feature on in `staging` so people can test it without running Calypso locally. After that we'll enable it for `production`, wait a couple of days more for regressions, and after that we'll be ready to release a version of the plugin that uses the new flow.

How does that sound? I just made it up, there was no real "plan" until now :)

**Comment by DanReyLop** [comment]
> are we recording tracks events separately for this flow?

No. I kept all the `tracks` events intact, but I'm not sure if there are any events that include the `from` parameter so we can detect if it's the new flow.

> should we start a new issue for the Jetpack-branded screen showing up, or is that going to be solved as part of this PR?

Neither. It's a simple fix, but it's on the wp.com server side, so that will be a diff on Phabricator. I'll cc/ you :)

> for the "to do" items mentioned in the PR description (passwordless flow, etc) - should we record those tasks in a separate issue to make sure we can keep track?

Yes, I'll open new issues with the remaining work before merging this one.

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:126]: When you put an email/username in the initial "login" form, a request is made to the server to see what kind of account that username/email has. The possible answers can be:
- Doesn't exist
- Regular account
- Passwordless account

Depending on that answer, the next step of the form will be different.
- If there is no account, the signup form will render.
- If it's a regular account, the password field will appear.
- If it's a passwordless account, a notice telling the user to check their e-mail will appear (I haven't tested this one out yet).

In this case, because the user is going back to the initial "login" form, we want to clear that response so they can input another username/email.

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:401]: See my reply [here](https://github.com/Automattic/wp-calypso/pull/41798#discussion_r432938870). Basically, the initial state is:
- Account type is not set (the user hasn't submitted a username/email to check its status)
- The only field visible is the Username/email field (as shown in the designs).

Then, when the user inputs an email and clicks the submit button, a request is made to the server. If there's an account with that email, the state will be:
- Account type will be set, it will be either `regular` or `passwordless` (not sure if these are the actual values, but that's not important right now).
- As such, at this point the user has the username/email field prefilled, and the rest of the login form (the password field) must appear.

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js]: I thought it made the code a bit more readable, but that's dubious. I've just changed it in https://github.com/Automattic/wp-calypso/pull/41798/commits/a49c0296104aa552dd3552be175f2eee2612287d, is that better or worse?

File: client/layout/logged-out.jsx
Comment (DanReyLop) [reply client/layout/logged-out.jsx]: Sure. It's used as a bolean, so it doesn't hurt to make it explicit. Done in https://github.com/Automattic/wp-calypso/pull/41798/commits/a9861190a0922bb53497ea7e0fde3565fe7efd2d

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js]: I've changed it in https://github.com/Automattic/wp-calypso/pull/41798/commits/433900c08ffa2a00f80addc7b9be9fb3921470c0, I think it reads a bit better now :)

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:110]: It just shows an error:
Screenshot 2020-05-31 at 12 55 02

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:290]: > This is basically the difference between the "enter your email" screen and the "log in" screen right? 

Yes.

> change the variable name to something like `! isLoggedOutScreen`.

Both screens are "logged out" screens. Really, both screens are part of the "login" flow, just different steps on it. I understand that `isFullLoginFormVisible` is not the perfect variable name but it's the best I could come up with. If you can't think of anything better, I'll add a clarifying comment :)

> Or we could designate three states in the component local state for the type of Woo DNA screen we're on

I'm not a fan of this option. The `isFullLoginFormVisible` prop is derived from the global Redux state, so we shouldn't store its value in the local component state.

File: client/jetpack-connect/authorize.js
Comment (DanReyLop) [reply client/jetpack-connect/authorize.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/41798/commits/e75fab9377165f088ad48456aa7828182f562366

File: client/blocks/login/login-form.jsx
Comment (DanReyLop) [reply client/blocks/login/login-form.jsx:486]: > Is DNA an official name that's being used?

No, it's not official, I made it up. I kind-of combined "Woo" with "Jetpack DNA" (which was the previous name of the Jetpack packaging efforts). The new flow didn't have a name, so i just used "Woo DNA" as a codename everywhere. Same way as "Gutenboarding" or "Frankenflow" (it's real, look it up), I expect that it will be just an internal name, and it being unique enough so it's easy to search for.

I'm not married to the "Woo DNA" name but I think it sounds cool :)

As to why it's `jetpackWooDnaFlow` instead of `wpcomWooDnaFlow` or something like that, Jetpack should still be part of the name. This file in particular, for example (`login-form.jsx`) is not specific to Jetpack, so being explicit that we're dealing with a Jetpack-related flow is important. Even if it's not Jetpack-branded, it's still a Jetpack connection flow internally, so the code shouldn't be trying to obfuscate that.

File: client/jetpack-connect/woo-dna-config.js
Comment (DanReyLop) [inline client/jetpack-connect/woo-dna-config.js:10]: @allendav FYI, we should have this page ready before releasing WCPay 1.1 (which will include the new connection flow).

Comment (allendav) [reply client/jetpack-connect/woo-dna-config.js:10]: OK - at that URL, right?  Please open an issue (type: task), start a google doc where we can collaborate on the content, and add the issue to the sprint 14 pre-planning post

Comment (DanReyLop) [reply client/jetpack-connect/woo-dna-config.js:10]: No, that URL will need to be changed, right now it points to our WCPay docs. It will need to be a different page or an "anchored" section on that page. I've created https://github.com/Automattic/woocommerce-payments/issues/708 and opened a doc to collaborate on this.

File: config/development.json
Comment (vbelolapotkov) [reply config/development.json:85]: @DanReyLop I like the plan but I expect we need someone from calypso team to review and approve it.

File: client/state/login/reducer.js
Comment (DanReyLop) [inline client/state/login/reducer.js:87]: This is so the user gets dropped in the same URL after logging in, instead of defaulting to `/`.

File: client/jetpack-connect/signup.js
Comment (vbelolapotkov) [inline client/jetpack-connect/signup.js:418]: `requestError` looks like too generic name for the prop. How about `loginRequestError` or something even more specific?

**Review by vbelolapotkov — COMMENTED** [review]
@DanReyLop well done! I've tested it locally and carefully looked through the code, added some comments and thoughts. None of them super critical but would be nice to hear your thoughts.

FYI, I haven't checked 2FA/Socials flow yet, will check them in #43059

File: client/components/jetpack-header/index.jsx
Comment (jsnajdr) [reply client/components/jetpack-header/index.jsx:55]: The best approach would be to rewrite `JetpackHeader` as function component and use the `useTranslate()` hook. That's more streamlined than the `localize` HOC.

If we don't do that, using the `localize` HOC is still the second best solution. Passing the `translate` prop from above is a bit unwieldy (it's not guaranteed that the parent component uses it itself), and ignoring i18n is also a :-1:

**Comment by DanReyLop** [comment]
> FYI, I haven't checked 2FA/Socials flow yet, will check them in #43059

You should still test them in this branch. I rebased so it's on top of everything. #43059 is a needed refactor, and the actual fix concerning the Woo DNS flow is in https://github.com/Automattic/wp-calypso/pull/41798/commits/1aebc79bc32e3de59174e88c633964a58dfe1c6f

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:349]: No, because then `showWooDnaLoginView` would be called with `event` as the first parameter, and [that would be wrong](https://github.com/Automattic/wp-calypso/blob/1aebc79bc32e3de59174e88c633964a58dfe1c6f/client/jetpack-connect/signup.js#L121).

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:116]: There's no need, [these other state values](https://github.com/Automattic/wp-calypso/blob/1aebc79bc32e3de59174e88c633964a58dfe1c6f/client/jetpack-connect/signup.js#L125-L126) are only used in the login view, so there's no need to clear them here. And we definitely don't want to clear the `signUpUsernameOrEmail` value, since that will be use to prefill the signup form.

File: client/jetpack-connect/signup.js
Comment (vbelolapotkov) [inline client/jetpack-connect/signup.js:328]: As I understand this is our starting point when user is not logged in. But it does not allow to login with socials and if socials associated email is entered the user is redirected to passwordless flow (see gif below). IMO we should have the view similar to the [wp.com sign-in](https://wordpress.com/log-in) with input field and socials options. Note there is a message regarding new account created in case of first time login with socials. I know it's not by flow design, so cc @LevinMedia for the design feedback.

woo_dna_google_login

Comment (vbelolapotkov) [reply client/jetpack-connect/signup.js:328]: Extra note here at the first step (when email or username input is shown alone):

* If I add non-existing email or userId and try to use social buttons I receive error https://cloudup.com/cXPGtrSWpIv.
* If I add existing email (not matching my social email) and then use social buttons I successfully forwarded to KYC flow.

**Comment by vbelolapotkov** [comment]
FYI, SMS 2FA flow is not fully branded but works.
image

**Comment by DanReyLop** [comment]
> FYI, SMS 2FA flow is not fully branded but works

What do you mean? The green checkmark is "success green", not "Jetpack green" :)

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:328]: This is becoming quite a rabbit-hole :)

Let's address the passwordless and Social login in a different branch. This PR is already massive, so as long as these changes form a cohesive experience and they don't break the existing flows, I think we should merge it sooner rather than later.

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:386]: > it will be hard to extract shared logic

There is a lot of shared logic between those 2 "components". They're just doing the same tasks in slightly different ways. I think it would be counter-productive to separate them because then we would need to keep them "in sync", which won't happen.

File: client/jetpack-connect/woo-dna-config.js
Comment (DanReyLop) [reply client/jetpack-connect/woo-dna-config.js:9]: I agree. However, this will change before release (sorry for this last-minute change, can be done in a different PR). The plugin name and help URL will come as part of the URL (the logic to "inject them" in the URL will be in the WPCom server), so there's no point in refactoring this.

File: client/jetpack-connect/auth-form-header.jsx
Comment (DanReyLop) [reply client/jetpack-connect/auth-form-header.jsx]: Done in https://github.com/Automattic/wp-calypso/pull/41798/commits/c865aa7e47e0d9442ffb72303a85ccccfef2bef1

File: client/jetpack-connect/authorize.js
Comment (DanReyLop) [reply client/jetpack-connect/authorize.js]: Done in https://github.com/Automattic/wp-calypso/pull/41798/commits/00cd893dc40ea8ef430f2f671e7be2ae23057681

File: client/jetpack-connect/authorize.js
Comment (DanReyLop) [reply client/jetpack-connect/authorize.js]: Done in https://github.com/Automattic/wp-calypso/pull/41798/commits/7f98ca56e6c3826350573242df68a2b1f559bff8

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js]: Done in https://github.com/Automattic/wp-calypso/pull/41798/commits/a7dc628ea10846b4bea6e5b0b4887b1b74ed9fb5

File: client/jetpack-connect/authorize.js
Comment (DanReyLop) [reply client/jetpack-connect/authorize.js:75]: See [this comment](https://github.com/Automattic/wp-calypso/pull/41798#discussion_r437736627). Hopefully I'll come up with a better name when I refactor it. Do you have any ideas? Renaming is easy :)

**Comment by vbelolapotkov** [comment]
> What do you mean? The green checkmark is "success green", not "Jetpack green" :)

Ah, sorry it looks so close to Jetpack login styling (green color, black bg) that I thought it's left unnoticed 🙂. Checked WordPress.com and it's the same there :) 

Noticed the other little thing though, `share details` in the approve page message 👇 is a link which is not obvious. Only cursor changes on hover. Should we add at least `text-decoration: underline` to it?

image

**Review by vbelolapotkov — APPROVED** [review]
LGTM! IMO it's fine to move it forward according to [shipping plan](https://github.com/Automattic/wp-calypso/pull/41798#discussion_r432864916).

Pre-approving with two minor fixes left:
1. [This comment](https://github.com/Automattic/wp-calypso/pull/41798#discussion_r437277910).
2 And [this one](https://github.com/Automattic/wp-calypso/pull/41798#issuecomment-641786254). This one can be fixed later though.

**Review by jsnajdr — APPROVED** [review]
> Added a review label back because final review from @Automattic/team-calypso is needed.

My concerns have been addressed, but I'm not really an expert on Jetpack Connect and I didn't test the flow. Therefore, a :+1: from me as a member of Team Calypso doesn't carry that much weight 🙂

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:389]: > As for keeping it in sync we have to do it anyway for render and renderWooDna methods.

Yeah, I was more worried about duplicating the rest of the logic. But if it can be extracted with react hooks, it should be possible to keep it relatively clean.

> That's why I think it won't be a simple refactoring and should be handled separately.

I've created https://github.com/Automattic/wp-calypso/issues/43187 to tackle this after the deadline.

File: client/jetpack-connect/woo-dna-config.js
Comment (DanReyLop) [reply client/jetpack-connect/woo-dna-config.js:9]: No, no issue, I'll open the PR (and the corresponding diff in wpcom) this evening or tomorrow.

**Comment by DanReyLop** [comment]
> Noticed the other little thing though, share details in the approve page message 👇 is a link which is not obvious.

It's so "not obvious" that I just realized there's a link there at all! It seems that styling was a deliberate choice (https://github.com/Automattic/wp-calypso/pull/34414), but anyway, I've changed it in https://github.com/Automattic/wp-calypso/pull/41798/commits/2f748bad38c1df6ace8b5dd16445bd5bf0424f9f:
Screenshot 2020-06-11 at 16 21 59

I hope not many people click that link though, it links to [a Jetpack support page](https://jetpack.com/support/what-data-does-jetpack-sync/).

**Comment by DanReyLop** [comment]
@vbelolapotkov On second thought, let's keep the link a bit "hidden". For WCPay users, clicking that link will just confuse them. I've tweaked the style again in https://github.com/Automattic/wp-calypso/pull/41798/commits/65ee0f2e015590d586df60b1d814b92320f6cd58

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3817309

Hi @DanReyLop, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Get help setting up %(pluginName)s`
- `Create a new account or connect as a different user`
- `Your account will enable you to start using the features and benefits offered by %(pluginName)s`
- `Enter your email address to get started`
- `Approve your connection`
- `WooCommerce logo`

Thank you in advance!

File: client/jetpack-connect/signup.js
Comment (DanReyLop) [reply client/jetpack-connect/signup.js:328]: > If I add non-existing email or userId and try to use social buttons I receive error https://cloudup.com/cXPGtrSWpIv.

I've fixed this on https://github.com/Automattic/wp-calypso/pull/43269

> if socials associated email is entered the user is redirected to passwordless flow (see gif below).

I'll work on this next. It's not a deal-breaker because, even though the redirect to wp.com sign-in is annoying, the user also correctly receives an email with the login link, so they can finish the flow.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3817309).

**Pull request merged**
